From 7fa27e5c52c9b69e1602be5409a628a99b9b4942 Mon Sep 17 00:00:00 2001
From: Josua Mayer <josua@solid-run.com>
Date: Thu, 18 May 2023 17:12:46 +0300
Subject: [PATCH 29/30] re-add all the ethernet phy register writes

---
 .../solidrun/am64x_solidrun/am64x_solidrun.c  | 31 ++++++++++++++++++-
 1 file changed, 30 insertions(+), 1 deletion(-)

diff --git a/board/solidrun/am64x_solidrun/am64x_solidrun.c b/board/solidrun/am64x_solidrun/am64x_solidrun.c
index 0a64eaaec6..5be4dd129f 100644
--- a/board/solidrun/am64x_solidrun/am64x_solidrun.c
+++ b/board/solidrun/am64x_solidrun/am64x_solidrun.c
@@ -319,6 +319,8 @@ int board_phy_config(struct phy_device *phydev)
 	ret = phy_write(phydev, MDIO_DEVAD_NONE, AM64X_SR_PHY_GEN_CTRL_REG, 0x8000);
 	if (ret)
 		return ret;
+
+	udelay(30);
 #endif
 
 	/*
@@ -331,8 +333,32 @@ int board_phy_config(struct phy_device *phydev)
 	 * - GEN_CTRL = 0x4000
 	 */
 
+	/* write into BMCR
+	 * speed selection = 1000 Mbps
+	 * A/N enable
+	 * duplex = full
+	 */
+	ret = phy_write(phydev, MDIO_DEVAD_NONE, AM64X_SR_PHY_BMCR_REG, 0x1140);
+	if (ret)
+		return ret;
+
+
+	/* write into GEN_CFG1*/
+	ret = phy_write(phydev, MDIO_DEVAD_NONE, AM64X_SR_PHY_GEN_CFG1_REG, 0xb00);
+	if (ret)
+		return ret;
+
+	/* write into PHY_CONTROL */
+	ret = phy_write(phydev, MDIO_DEVAD_NONE, AM64X_SR_PHY_PHY_CONTROL_REG, 0x5048);
+	if (ret)
+		return ret;
+
+	/* write into OP_MODE_DECODE */
+	ret = am64x_sr_phy_write_mmd(phydev, AM64X_SR_PHY_OP_MODE_DECODE_REG, 0x0);
+	if (ret)
+		return ret;
+
 	/* configure clk-out */
-	printf("configuring phy clock output ...\n");
 	ret = am64x_sr_phy_write_mmd(phydev, AM64X_SR_PHY_IO_MUX_CFG_REG, 0x0C10);
 	if (ret)
 		return ret;
@@ -360,6 +386,9 @@ int board_phy_config(struct phy_device *phydev)
 	if (ret)
 		return ret;
 
+	/* write into GEN_CTRL - SW reset*/
+	ret = phy_write(phydev, MDIO_DEVAD_NONE, AM64X_SR_PHY_GEN_CTRL_REG, 0x4000);
+
 	return ret;
 }
 
-- 
2.35.3

